uxarray/grid.py

```diff
@@ -0,0 +1,272 @@
+            pass
+
+    # vertices init
+    def from_vert(self):
+        """Create a grid with one face with vertices specified by the given
+        argument."""
+        print("Initializing with vertices")
+        self._init_mesh2()
+        self.in_ds.Mesh2.attrs['topology_dimension'] = self.vertices[0].size
+
+        self.in_ds["Mesh2"].topology_dimension
+
+        x_coord = self.vertices.transpose()[0]
+        y_coord = self.vertices.transpose()[1]
+
+        # single face with all nodes
+        num_nodes = x_coord.size
+        conn = list(range(0, num_nodes))
+        conn = [conn]
+
+        self.in_ds["Mesh2_node_x"] = xr.DataArray(data=xr.DataArray(x_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_node_y"] = xr.DataArray(data=xr.DataArray(y_coord),
+                                                  dims=["nMesh2_node"])
+        self.in_ds["Mesh2_face_nodes"] = xr.DataArray(
+            data=xr.DataArray(conn),
+            dims=["nMesh2_face", "nMaxMesh2_face_nodes"],
+            attrs={
+                "cf_role": "face_node_connectivity",
+                "_FillValue": -1,
+                "start_index": 0
+            })
+
+    # TODO: gridspec init
+    def from_gridspec(self):
+        print("initializing with gridspec")
+
+    # load mesh from a file
+    def from_file(self):
+        """Loads a mesh file Also, called by __init__ routine This routine will
+        automatically detect if it is a UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+        # find the file type
+        self.meshFileType = self.find_type()
+
+        # call reader as per meshFileType
+        if self.meshFileType == "exo":
+            self.read_exodus(self.ext_ds)
+        elif self.meshFileType == "scrip":
+            self.read_scrip(self.ext_ds)
+        elif self.meshFileType == "ugrid":
+            self.read_ugrid(self.filepath)
+        elif self.meshFileType == "shp":
+            self.read_shpfile(self.filepath)
+        else:
+            raise RuntimeError("unknown file format.")
+
+    # helper function to find file type
+    def find_type(self):
+        """Checks file path and contents to determine file type Also, called by
+        __init__ routine This routine will automatically detect if it is a
+        UGrid, SCRIP, Exodus, or shape file.
+
+        Raises:
+            RuntimeError: Invalid file type
+        """
+
+        try:
+            # extract the file name and extension
+            path = PurePath(self.filepath)
+            file_extension = path.suffix
+
+            # try to open file with xarray
+            self.ext_ds = xr.open_dataset(self.filepath, mask_and_scale=False)
+            #
+        except (TypeError, AttributeError) as e:
+            msg = str(e) + ': {}'.format(self.filepath)
+            print(msg)
+            raise RuntimeError(msg)
+            exit
```

Comment:
I'm not sure this is proper implementation of python try catch blocks with raise exceptions. At the very least the exit call at the bottom of the blocks should be a function call.

```suggestion
        try:
            # extract the file name and extension
            path = PurePath(self.filepath)
            file_extension = path.suffix

            # try to open file with xarray
            self.ext_ds = xr.open_dataset(self.filepath, mask_and_scale=False)
            #
        except (TypeError, AttributeError) as e:
            msg = str(e) + ': {}'.format(self.filepath)
            print(msg)
            raise RuntimeError(msg)
            exit()
```